File: level3/ex04.py
# dp[i] = dp[k] (+, -, *, //) d[i - k] (0 < k < i), dp[i]를 구하는 함수
def getNumbers(dp, i):
    for k in range(1, i):
        for value1 in dp[k]:
            for value2 in dp[i - k]:
                # +, -, * 연산 결과
                result = [value1 + value2, value1 - value2, value1 * value2]
                # //
                if value2 != 0:
                    result.append(value1 // value2)
                # 음수 제거
                result = filter(lambda x: x >= 0, result)
                # dp[i]에 추가
                dp[i] = dp[i].union(set(result))

def solution(N, number):
    # dp[i]: N을 i번 사용해서 만들 수 있는 수 집합
    # dp[i] = dp[k] (+, -, *, //) d[i - k] (0 < k < i)
    dp = [set([0]), set([N])]

    for i in range(2, 9):
        # N이 i번 연속된 수 추가
        dp.append(set([N * int('1' * i)]))      
        # dp[i] = dp[k] (+, -, *, //) d[i - k] (0 < k < i)    
        getNumbers(dp, i)

    answer = -1

    for i in range(9):
        if number in dp[i]:
            answer = i
            break

    return answer

File: level3/ex04.js
function solution(N, number) {

    // dp[i]에 해당하는 사칙연산 결과값을 반환하는 함수
    // dp[i] = dp[k] (+, -, *, //) d[i-k] (0 < k < i)
    function getResult(dp, i) {
        const result = [];
        
        for (let k=1; k<i; k++) {
            for (const value1 of dp[k]) {
                for (const value2 of dp[i-k]) {
                    // +, -, * 결과값 추가
                    result.push(value1+value2);
                    result.push(value1-value2);
                    result.push(value1*value2);
                    // // 결과값 추가
                    if (value2 > 0) {
                        result.push(parseInt(value1/value2));
                    }
                    
                }
            }
        }
        
        // 음수 값 제거하여 반환
        return result.filter(value => value > 0);
    }

    // dp[i]: N을 i번 사용해서 만들 수 있는, 모든 수의 집합
    const dp = new Array(9).fill([]); // 최솟값이 8보다 크면 -1을 반환
    dp[0] = [0];
    dp[1] = [N];
    
    for (let i = 2; i <= 8; i++) {
        // N이 i번 연속된 수 추가
        dp[i].push(N.toString().repeat(i));
        // 그 외 사칙연산 결과값 추가 
        dp[i] = [
            parseInt(N.toString().repeat(i)),
            ...getResult(dp, i)
        ];
    } 
    
    let answer = -1;
    
    for (let i = 0; i <= 8; i++) {
        if (dp[i].includes(number)) {
            answer = i;
            break;
        }
    }
    
    return answer;
}
